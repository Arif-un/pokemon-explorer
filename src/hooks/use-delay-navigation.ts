import { useState } from 'react'

import { useRouter } from '@tanstack/react-router'

interface NextLocation {
  to: string
  replace: boolean
}

export function useDelayedNavigation() {
  const router = useRouter()
  const [isNavigating, setIsNavigating] = useState(false)
  const [nextLocation, setNextLocation] = useState<NextLocation | null>(null)

  const navigate = (to: string, { replace = false } = {}) => {
    setIsNavigating(true)
    setNextLocation({ to, replace })
  }

  const finishNavigation = () => {
    if (nextLocation) {
      const { to, replace } = nextLocation
      router.navigate({ to, replace, search: prv => ({ ...prv }) })
      setIsNavigating(false)
      setNextLocation(null)
    }
  }

  return { isNavigating, navigate, finishNavigation }
}
