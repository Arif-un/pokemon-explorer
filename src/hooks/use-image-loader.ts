import { useEffect, useState } from 'react'

interface ImageState {
  src: string
  isLoading: boolean
  error: Event | string | null
}

const imageCache: Record<string, boolean> = {}

export function useImageLoader(src: string, fallbackSrc: string): ImageState {
  const [imageState, setImageState] = useState<ImageState>({
    src: fallbackSrc,
    isLoading: true,
    error: null
  })

  useEffect(() => {
    if (imageCache[src]) {
      setImageState({
        src: src,
        isLoading: false,
        error: null
      })
      return
    }

    const img = new Image()
    img.src = src

    img.onload = () => {
      imageCache[src] = true

      setImageState({
        src: src,
        isLoading: false,
        error: null
      })
    }

    img.onerror = error => {
      setImageState({
        src: fallbackSrc,
        isLoading: false,
        error
      })
    }

    return () => {
      img.onload = null
      img.onerror = null
    }
  }, [src, fallbackSrc])

  return imageState
}
