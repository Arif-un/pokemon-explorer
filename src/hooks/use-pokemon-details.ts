import { useQuery, useQueryClient } from '@tanstack/react-query'

import { fetchPokemonDetails } from '@/services/pokemon-api'
import { type PokemonDetailsResponse, type PokemonResponse } from '@/types/Pokemon'
import { preparePokemonDetailsInitialData } from '@/utils/query-helpers'

export const usePokemonDetails = <SelectedData = PokemonDetailsResponse>(
  name: string,
  select: (data: PokemonDetailsResponse) => SelectedData
) => {
  const queryClient = useQueryClient()

  const { data, isLoading, isError, error } = useQuery<PokemonDetailsResponse, Error, SelectedData>({
    queryKey: ['pokemon', name],
    queryFn: () => fetchPokemonDetails(name),
    initialData: () => {
      const cachedPokemonQueries = queryClient.getQueriesData<PokemonResponse>({
        queryKey: ['pokemon']
      })
      return preparePokemonDetailsInitialData(name, cachedPokemonQueries)
    },
    select
  })

  return { data: data as SelectedData, isLoading, isError, error }
}
