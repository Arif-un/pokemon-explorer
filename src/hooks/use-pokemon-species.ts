import { useQuery } from '@tanstack/react-query'

import { fetchPokemonSpecies } from '@/services/pokemon-api'

export const usePokemonSpecies = (name: string) => {
  const { data, isLoading, isError, error } = useQuery({
    queryKey: ['pokemon-color', name],
    queryFn: () => fetchPokemonSpecies(name),
    staleTime: 5 * 60 * 1000
  })

  return {
    pokemonColor: data?.color.name,
    isLoading,
    isPokemonSpeciesError: isError,
    pokemonSpeciesError: error
  }
}
