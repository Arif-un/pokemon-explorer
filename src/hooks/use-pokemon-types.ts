import { useQuery } from '@tanstack/react-query'

import { fetchPokemonTypes } from '@/services/pokemon-api-graphql'
import { DEFAULT_QUERY_STALE_TIME } from '@/utils/constants'

export function usePokemonTypes() {
  const { data, isLoading, isError, error } = useQuery({
    queryKey: ['pokemon-types'],
    queryFn: fetchPokemonTypes,
    staleTime: DEFAULT_QUERY_STALE_TIME
  })

  return {
    types: data?.pokemon_v2_type.sort((a, b) => a.name.localeCompare(b.name)),
    isLoading,
    isError,
    error
  }
}
