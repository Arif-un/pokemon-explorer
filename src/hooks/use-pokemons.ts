import { useQuery } from '@tanstack/react-query'

import { fetchPokemonWithFilters } from '@/services/pokemon-api-graphql'
import {
  DEFAULT_POKEMON_FETCH_LIMIT,
  DEFAULT_POKEMON_FETCH_OFFSET,
  DEFAULT_QUERY_STALE_TIME
} from '@/utils/constants'

interface UsePokemonsParams {
  name?: string
  limit?: number
  offset?: number
  types?: string
  sort?: string
}

export const usePokemons = ({
  name,
  limit = DEFAULT_POKEMON_FETCH_LIMIT,
  offset = DEFAULT_POKEMON_FETCH_OFFSET,
  types,
  sort
}: UsePokemonsParams) => {
  const { data, isLoading, isError, error, isFetching } = useQuery({
    queryKey: ['pokemon', name, limit, offset, types, sort],
    queryFn: () => fetchPokemonWithFilters({ name, limit, offset, types, sort }),
    staleTime: DEFAULT_QUERY_STALE_TIME,
    placeholderData: previousData => previousData
  })

  return {
    pokemons: data?.pokemons,
    totalItems: data?.pokemon_aggregate.aggregate.count,
    isLoading,
    isError,
    error,
    isFetching
  }
}
