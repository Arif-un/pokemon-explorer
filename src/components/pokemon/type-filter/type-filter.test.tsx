import { useSearchParam } from 'react-use'

import { useNavigate } from '@tanstack/react-router'
import { render, screen } from '@testing-library/react'
import ResizeObserver from 'resize-observer-polyfill'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import { usePokemonTypes } from '@/hooks/use-pokemon-types'

import TypeFilter from './type-filter'

vi.mock('@/hooks/use-pokemon-types')
vi.mock('react-use')
vi.mock('@tanstack/react-router')

global.ResizeObserver = ResizeObserver

describe('TypeFilter', () => {
  const mockTypes = [{ name: 'fire' }, { name: 'water' }, { name: 'grass' }]

  const mockNavigate = vi.fn()

  beforeEach(() => {
    vi.mocked(usePokemonTypes).mockReturnValue({
      types: mockTypes,
      isLoading: false,
      error: null,
      isError: false
    })
    vi.mocked(useSearchParam).mockReturnValue('')
    vi.mocked(useNavigate).mockReturnValue(mockNavigate)
  })

  it('renders correctly', () => {
    const { asFragment } = render(<TypeFilter />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('displays the correct number of selected types', () => {
    vi.mocked(useSearchParam).mockReturnValue('fire-water')
    render(<TypeFilter />)
    expect(screen.getByDisplayValue('Filter by types (2)')).toBeInTheDocument()
  })

  it('handles loading state', () => {
    vi.mocked(usePokemonTypes).mockReturnValue({
      types: [],
      isLoading: true,
      error: null,
      isError: false
    })
    render(<TypeFilter />)
    expect(screen.getByRole('button')).toBeDisabled()
  })

  it('handles error state', () => {
    vi.mocked(usePokemonTypes).mockReturnValue({
      types: [],
      isLoading: false,
      error: 'Error' as unknown as Error,
      isError: true
    })
    render(<TypeFilter />)
    expect(screen.getByRole('button')).toBeDisabled()
  })
})
