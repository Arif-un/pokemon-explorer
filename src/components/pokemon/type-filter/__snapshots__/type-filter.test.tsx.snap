// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TypeFilter > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="w-44 inline-block"
  >
    <div
      class="relative"
    >
      <input
        aria-autocomplete="list"
        aria-expanded="false"
        aria-label="Type filter"
        class="w-full rounded-full border py-1.5 pr-8 pl-3 h-11 text-sm/6 dark:text-white dark:border-slate-800 border-slate-300 bg-slate-50 dark:bg-slate-900 select-none focus:outline-none data-[focus]:outline-2 data-[focus]:-outline-offset-1 data-[focus]:outline-yellow-500"
        data-headlessui-state=""
        id="headlessui-combobox-input-:r2:"
        readonly=""
        role="combobox"
        type="text"
        value=""
      />
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-label="Open type filter menu"
        class="group rounded-full w-full h-11 flex justify-end items-center dark:text-slate-400 inset-y-0 my-auto absolute px-3"
        data-headlessui-state=""
        id="headlessui-combobox-button-:r3:"
        tabindex="-1"
        type="button"
      >
        <svg
          class="lucide lucide-chevron-down size-4 dark:text-slate-50"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m6 9 6 6 6-6"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
